pwiz_tools/Skyline/Model/AuditLog/AuditLogEntry.cs

```diff
@@ -300,24 +328,38 @@ public static bool ReadFromXmlTextReader(XmlTextReader reader, out string logged
                 }
             }
 
-            string rootHash = null;
+            string rootHashString = null;
             if (reader.IsStartElement(EL.root_hash.ToString()))
-                rootHash = reader.ReadElementString(EL.root_hash.ToString());
+                rootHashString = reader.ReadElementString(EL.root_hash.ToString());
 
             result = reader.DeserializeElement<AuditLogList>();
             result.FormatVersion = docFormat;
             result.RootHash = new AuditLogHash()
                 .ChangeActualHash(result.CalculateRootHash())
-                .ChangeSkylHash(Hash.FromBase64(rootHash));
+                .ChangeSkylHash(Hash.FromBase64(rootHashString));
 
-            // yell if audit log entry hash does not match.
-            // If the docFormat is null, this is an old audit log and there won't be any entry hashes.
-            // We can't just always ignore non-existent hashes, otherwise people could just delete the hash elements
+            // Root and entry-level hashes validation
+            // We can't ignore non-existent hashes, otherwise people could just delete the hash elements
             // and get Skyline to successfully load the audit log
-            var modifiedEntries =
-                result.AuditLogEntries.Enumerate().Where(entry => !entry.Hash.SkylAndActualHashesEqual()).ToArray();
 
-            if (docFormat != null && (!result.RootHash.SkylAndActualHashesEqual() || modifiedEntries.Length > 0))
+            // recalculate the hashes for the current document format
+            var hashes = result.AuditLogEntries.Enumerate().Select(e =>
```

Comment:
I think this would be easier to debug if you were calculating the hash and checking whether it was valid in the same pass through the loop.

So, you would have:

            var hashes = new List<AuditLogHash>();
            var entriesWithIncorrectHash = new List<AuditLogEntry>();
            foreach (var entry in result.AuditLogEntries.Enumerate())
            {
                AuditLogHash hash;
                if (DocumentFormat.CURRENT.Equals(docFormat))
                    hash = entry.Hash;
                else
                    hash = new AuditLogHash(entry, entry.Hash.SkylHash, docFormat);
                if (!hash.SkylAndActualHashesEqual())
                {
                    entriesWithIncorrectHash.Add(entry);
                }
                hashes.Add(hash);
            }

Reply:
Code updated.